Reordered and extended grammar documentation (#23)
langium/langium-website: content/docs/grammar-language.md

```
The parser rule `Person` will now lead to the creation of objects of type `OtherType` instead of `Person`.

### The Entry Rule
The *entry rule* is a parser rule that defines the starting point of the parsing step. The *entry rule* starts with the keyword `entry` and matches other parser rules.
```
entry Model:
    (persons+=Person | greetings+=Greeting)*;
```

> In this example, the *entry rule* `Model` defines a group of alternatives. The parser will go through the input document and try to parse `Person` or `Greeting` objects and add them to the `persons` or `greetings` arrays, respectively. The parser reads the token stream until all inputs have been consumed.

Reviewer: Am I correct that the parser reads all inputs until the end of the document **because** of the cardinality * ?

Wouldn't the parser stop after the first match if we do not have the cardinality * ?

If so then I think it would be good to mention it here.

```suggestion
In this example, the *entry rule* `Model` defines a group of alternatives. The parser will go through the input document and try to parse `Person` or `Greeting` objects and add them to the `persons` or `greetings` arrays, respectively. The parser reads the token stream until all inputs have been consumed thanks to the cardinality `*`.
```

Author: The parser always tries to consume the whole token stream. In case there are syntax errors, it may stop earlier (depending on the error recovery capabilities), but that's independent of the cardinality.

Author: Of course it's a syntax error to write more than one person or greeting if the `* ` is omitted, but the parser will still read the superfluous tokens.

### Extended Backus-Naur Form Expressions
Parser rules are defined using *Extended Backus-Naur Form*-like (EBNF) expressions similar to the [Xtext](https://www.eclipse.org/Xtext/) notation.

#### Cardinalities
A cardinality defines the number of elements in a given set. Four different cardinalities can be defined for any expression:
1. exactly one (no operator)
2. zero or one (operator `?`)
3. zero or many (operator `*`)
4. one or many (operator `+`)

#### Groups
Expressions can be put in sequence specifying the order they have to appear:
```
Person:
    'person' name=ID address=Address;
```
In this example, the rule `Person` must start with the `person` keyword followed by an `ID` token and an instance of the `Address` rule.

#### Alternatives

> It is possible to match one of multiple valid options by using the pipe operator `|`. The already mentioned `Model` example specifies to parse either `Person` or `Greeting` inside the parentheses:

Reviewer: To me it reads like an alternative will always be in parentheses.

```suggestion
It is possible to match one of multiple valid options by using the pipe operator `|`. The already mentioned `Model` example specifies to parse either `Person` or `Greeting`, zero or many times (cardinality *):
```

```
entry Model:
    (persons+=Person | greetings+=Greeting)*;
```

#### Keywords
Keywords are *inline terminals* which need to match a character sequence surrounded by single or double quotes, for example `'person'` or `"person"`. Keywords must not be empty and must not contain white space.

#### Assignments
Assignments define properties on the type returned by the surrounding rule.
There are three different ways to assign an expression (right side) to a property (left side).

1. `=` is used for assigning **a single value** to a property.
    ```
    Person:
        'person' name=ID
    ```
    Here, the property `name` will accept only one expression matching the terminal rule `ID`.

2. `+=` is used to assign **multiple values** to an array property.
    ```
    Contact:
        addresses+=STRING addresses+=STRING;
    ```
    Here, the array property `addresses` will accept two expressions matching the terminal rule `STRING`.

3. `?=` is used to assign a **value to a property of type boolean**. The value of the property of type `boolean` is set to `true` if the right part of the assignment is consumed by the parser.
    ```
    Employee:

>         'employee' name=ID (remote?='remote')?

Reviewer: Is there a difference between `(remote?='remote')?` and `remote?='remote'?` ?

They both seem to do what I expect in the LSP, is there some differences to be expected in the AST/CST ?

Author: No it's exactly the same, so the parentheses are not necessary here.

#### Unassigned Rule Calls
Parser rules do not necessarily need to create an object, they can also refer to other parser rules which in turn will be responsible for returning the object.
For example, in the [Arithmetics example](https://github.com/langium/langium/blob/main/examples/arithmetics/src/language-server/arithmetics.langium):
```
AbstractDefinition:
	Definition | DeclaredParameter;
```
The parser rule `AbstractDefinition` will not create an object of type AbstractDefinition. Instead, it calls either the `Definition` or `DeclaredParameter` parser rule which will be responsible for creating an object of a given type (or call other parser rules if they are unassigned rule calls themselves).

In contrast, an assigned rule call such as `parameter=DeclaredParameter` means that an object is created in the current parser rule and assigns the reult of the `DeclaredParameter` parser rule to the specified property `parameter` of that object.

#### Unordered Groups
*Unordered groups are not supported in the current version of Langium. Please refer to [this issue](https://github.com/langium/langium/issues/314) for further information.*

> In regular groups, expressions must occur in the exact order as they are declared.

Reviewer: ```suggestion
In regular groups, expressions must occur in the exact order they are declared.
```
